Assets/MRTK/SDK/Experimental/Joystick/JoystickController.cs

```diff
@@ -0,0 +1,190 @@
+            // Forward Back = Vertical
+            joystickVisualRotation.x = Mathf.Clamp(joystickGrabberPosition.z * SensitivityForwardBack,-joystickVisualMaxRotation, joystickVisualMaxRotation);
+            // TODO: calculate joystickVisualRotation.y to always face the proper direction (for when the joystick container gets moved around the scene)
+            if (joystickVisual != null)
+            {
+                joystickVisual.transform.localRotation = Quaternion.Euler(joystickVisualRotation);
+            }
+        }
+
+        private void ApplyJoystickValues()
+        {
+            if(ObjectToManipulate != null)
+            {
+                if (Mode == JoystickMode.Move)
+                {
+                    ObjectToManipulate.transform.position += (joystickGrabberPosition * MoveSpeed);
+                    if (debugText != null)
+                    {
+                        debugText.text = ObjectToManipulate.transform.position.ToString();
+                    }
+                }
+                else if (Mode == JoystickMode.Rotate)
+                {
+                    Vector3 newRotation = ObjectToManipulate.transform.rotation.eulerAngles;
+                    // only take the horizontal axis from the joystick
+                    newRotation.y += (joystickGrabberPosition.x * RotationSpeed);
+                    newRotation.x = 0;
+                    newRotation.z = 0;
+                    ObjectToManipulate.transform.localRotation = Quaternion.Euler(newRotation);
+                    if (debugText != null)
+                    {
+                        debugText.text = ObjectToManipulate.transform.localRotation.eulerAngles.ToString();
+                    }
+                }
+                else if (Mode == JoystickMode.Scale)
+                {
+                    // TODO: Clamp above zero
+                    Vector3 newScale = new Vector3(joystickGrabberPosition.x, joystickGrabberPosition.x, joystickGrabberPosition.x) * ScaleSpeed;
+                    ObjectToManipulate.transform.localScale += newScale;
+                    if (debugText != null)
+                    {
+                        debugText.text = ObjectToManipulate.transform.localScale.ToString();
+                    }
+                }
+            }
+        }
+        /// <summary>
+        /// The ObjectManipulator script uses this to determine when the joystick is grabbed.
+        /// </summary>
+        public void StartDrag()
+        {
+            isDragging = true;
+        }
+        /// <summary>
+        /// The ObjectManipulator script uses this to determine when the joystick is released.
+        /// </summary>
+        public void StopDrag()
+        {
+            isDragging = false;
+        }
+        /// <summary>
+        /// Set the joystick mode from a UI button.
+        /// </summary>
+        public void JoystickModeMove()
```

Comment:
Are these functions because  you want to be able to access them via Unity's UI (i.e. hooking up events and stuff?)

Otherwise it would be a little odd because these are just setting an property (and a caller could easily do this because Mode is already public)

Reply:
You are correct. Per the summary, the UI for changing modes with buttons requires explicit functions. For example, on Acme, we have one joystick and 3 buttons that explicitly change the mode.
![alignment](https://user-images.githubusercontent.com/43013191/84526668-a31cd580-ac92-11ea-9cd9-5d40c9115343.png)